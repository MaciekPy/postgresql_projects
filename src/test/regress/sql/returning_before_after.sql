--
-- Test BEFORE/AFTER feature in RETURNING statements

CREATE TABLE foo_ret (
		bar1 INTEGER,
		bar2 TEXT
		);

INSERT INTO foo_ret VALUES (1, 'x'),(2,'y');

UPDATE foo_ret SET bar1=bar1+1 RETURNING before.*, bar1, bar2;

UPDATE foo_ret SET bar1=bar1-1 RETURNING after.bar1, before.bar1*2;

UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*;

-- check single after

UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'a' RETURNING after.*;

-- check single before

UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'b' RETURNING before.*;

-- it should fail
UPDATE foo_ret SET bar1=bar1+before.bar1 RETURNING before.*;
UPDATE foo_ret SET bar1=bar1+after.bar1 RETURNING after.*;

-- test before/after aliases
UPDATE foo_ret AS before SET bar1=bar1+1 RETURNING before.*,after.*;
UPDATE foo_ret AS after SET bar1=bar1-1 RETURNING before.*,after.*;

-- test inheritance
CREATE TABLE foo_ret2_ret (bar INTEGER) INHERITS(foo_ret);

INSERT INTO foo_ret2_ret VALUES (1,'b',5);

UPDATE foo_ret2_ret SET bar1=bar1*2, bar=bar1+5, bar2=bar1::text || bar::text RETURNING before.*, after.*, *;
UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*;

-- check views

CREATE VIEW view_foo_ret AS SELECT * FROM foo_ret;

UPDATE view_foo_ret SET bar1=bar1+1 RETURNING before.*, bar1, bar2;

CREATE TABLE foo_ret3 (bar1 INTEGER, bar4 FLOAT);

INSERT INTO foo_ret2_ret VALUES (2, 'asdf', 33);
INSERT INTO foo_ret3 VALUES (2, 7.77);

CREATE VIEW view_join AS SELECT f2.*, f3.bar1 AS f1bar1, f3.bar4 FROM foo_ret2_ret f2
JOIN foo_ret3 f3 ON f2.bar1 = f3.bar1;

UPDATE view_join SET bar1=bar1+5, bar2=bar2||'join', bar=bar1*2, bar4=7 RETURNING before.*, after.*;

-- check triggers
CREATE FUNCTION returning_trig() returns trigger as $$
BEGIN
NEW.bar1 = NEW.bar1*NEW.bar1;
RETURN NEW;
END; $$ language plpgsql;

DROP TABLE foo_ret2_ret CASCADE;
CREATE TRIGGER bef_foo_ret BEFORE UPDATE ON foo_ret FOR EACH ROW EXECUTE PROCEDURE returning_trig();

UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*, *;

DROP TABLE foo_ret CASCADE;
DROP TABLE foo_ret3 CASCADE;

CREATE TABLE t1_ret (id serial, x int, y int, z int);
CREATE TABLE t2_ret (id serial, x int, y int, z int);

INSERT INTO t1_ret VALUES (DEFAULT,1,2,3);
INSERT INTO t1_ret VALUES (DEFAULT,4,5,6);

-- check WITH statement
WITH foo_ret AS (UPDATE t1_ret SET x=x*2, y=y+1, z=x+y+z RETURNING BEFORE.x, BEFORE.y, AFTER.z) INSERT INTO t2_ret (x,y,z) SELECT x, y, z FROM foo_ret RETURNING *;

-- check UPDATE ... FROM statement
UPDATE t2_ret SET x = t1_ret.x+2 FROM t1_ret WHERE t2_ret.id=t1_ret.id RETURNING after.x, before.x;
UPDATE t2_ret SET x = t1_ret.x*2 FROM t1_ret WHERE t2_ret.id=t1_ret.id RETURNING after.*, before.*;

DROP TABLE t1_ret;
DROP TABLE t2_ret;
