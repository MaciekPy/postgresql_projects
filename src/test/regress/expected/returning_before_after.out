--
-- Test BEFORE/AFTER feature in RETURNING statements
CREATE TABLE foo_ret (
		bar1 INTEGER,
		bar2 TEXT
		);
INSERT INTO foo_ret VALUES (1, 'x'),(2,'y');
UPDATE foo_ret SET bar1=bar1+1 RETURNING before.*, bar1, bar2;
 bar1 | bar2 | bar1 | bar2 
------+------+------+------
    1 | x    |    2 | x
    2 | y    |    3 | y
(2 rows)

UPDATE foo_ret SET bar1=bar1-1 RETURNING after.bar1, before.bar1*2;
 bar1 | ?column? 
------+----------
    1 |        4
    2 |        6
(2 rows)

UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*;
 bar1 | bar2 | bar1 | bar2 
------+------+------+------
    1 | x    |    2 | xz
    2 | y    |    3 | yz
(2 rows)

-- check single after
UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'a' RETURNING after.*;
 bar1 | bar2 
------+------
    3 | xza
    4 | yza
(2 rows)

-- check single before
UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'b' RETURNING before.*;
 bar1 | bar2 
------+------
    3 | xza
    4 | yza
(2 rows)

-- it should fail
UPDATE foo_ret SET bar1=bar1+before.bar1 RETURNING before.*;
ERROR:  missing FROM-clause entry for table "before"
LINE 1: UPDATE foo_ret SET bar1=bar1+before.bar1 RETURNING before.*;
                                     ^
UPDATE foo_ret SET bar1=bar1+after.bar1 RETURNING after.*;
ERROR:  missing FROM-clause entry for table "after"
LINE 1: UPDATE foo_ret SET bar1=bar1+after.bar1 RETURNING after.*;
                                     ^
-- test before/after aliases
UPDATE foo_ret AS before SET bar1=bar1+1 RETURNING before.*,after.*;
 bar1 | bar2 | bar1 | bar2 
------+------+------+------
    5 | xzab |    5 | xzab
    6 | yzab |    6 | yzab
(2 rows)

UPDATE foo_ret AS after SET bar1=bar1-1 RETURNING before.*,after.*;
 bar1 | bar2 | bar1 | bar2 
------+------+------+------
    5 | xzab |    4 | xzab
    6 | yzab |    5 | yzab
(2 rows)

-- test inheritance
CREATE TABLE foo_ret2_ret (bar INTEGER) INHERITS(foo_ret);
INSERT INTO foo_ret2_ret VALUES (1,'b',5);
UPDATE foo_ret2_ret SET bar1=bar1*2, bar=bar1+5, bar2=bar1::text || bar::text RETURNING before.*, after.*, *;
 bar1 | bar2 | bar | bar1 | bar2 | bar | bar1 | bar2 | bar 
------+------+-----+------+------+-----+------+------+-----
    1 | b    |   5 |    2 | 15   |   6 |    2 | 15   |   6
(1 row)

UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*;
 bar1 | bar2 | bar1 | bar2  
------+------+------+-------
    4 | xzab |    5 | xzabz
    5 | yzab |    6 | yzabz
    2 | 15   |    3 | 15z
(3 rows)

-- check views
CREATE VIEW view_foo_ret AS SELECT * FROM foo_ret;
UPDATE view_foo_ret SET bar1=bar1+1 RETURNING before.*, bar1, bar2;
 bar1 | bar2  | bar1 | bar2  
------+-------+------+-------
    5 | xzabz |    6 | xzabz
    6 | yzabz |    7 | yzabz
    3 | 15z   |    4 | 15z
(3 rows)

CREATE TABLE foo_ret3 (bar1 INTEGER, bar4 FLOAT);
INSERT INTO foo_ret2_ret VALUES (2, 'asdf', 33);
INSERT INTO foo_ret3 VALUES (2, 7.77);
CREATE VIEW view_join AS SELECT f2.*, f3.bar1 AS f1bar1, f3.bar4 FROM foo_ret2_ret f2
JOIN foo_ret3 f3 ON f2.bar1 = f3.bar1;
UPDATE view_join SET bar1=bar1+5, bar2=bar2||'join', bar=bar1*2, bar4=7 RETURNING before.*, after.*;
ERROR:  cannot update view "view_join"
DETAIL:  Views that do not select from a single table or view are not automatically updatable.
HINT:  To enable updating the view, provide an INSTEAD OF UPDATE trigger or an unconditional ON UPDATE DO INSTEAD rule.
-- check triggers
CREATE FUNCTION returning_trig() returns trigger as $$
BEGIN
NEW.bar1 = NEW.bar1*NEW.bar1;
RETURN NEW;
END; $$ language plpgsql;
DROP TABLE foo_ret2_ret CASCADE;
NOTICE:  drop cascades to view view_join
CREATE TRIGGER bef_foo_ret BEFORE UPDATE ON foo_ret FOR EACH ROW EXECUTE PROCEDURE returning_trig();
UPDATE foo_ret SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*, *;
 bar1 | bar2  | bar1 |  bar2  | bar1 |  bar2  
------+-------+------+--------+------+--------
    6 | xzabz |    7 | xzabzz |   49 | xzabzz
    7 | yzabz |    8 | yzabzz |   64 | yzabzz
(2 rows)

DROP TABLE foo_ret CASCADE;
NOTICE:  drop cascades to view view_foo_ret
DROP TABLE foo_ret3 CASCADE;
CREATE TABLE t1_ret (id serial, x int, y int, z int);
CREATE TABLE t2_ret (id serial, x int, y int, z int);
INSERT INTO t1_ret VALUES (DEFAULT,1,2,3);
INSERT INTO t1_ret VALUES (DEFAULT,4,5,6);
-- check WITH statement
WITH foo_ret AS (UPDATE t1_ret SET x=x*2, y=y+1, z=x+y+z RETURNING BEFORE.x, BEFORE.y, AFTER.z) INSERT INTO t2_ret (x,y,z) SELECT x, y, z FROM foo_ret RETURNING *;
 id | x | y | z  
----+---+---+----
  1 | 1 | 2 |  6
  2 | 4 | 5 | 15
(2 rows)

-- check UPDATE ... FROM statement
UPDATE t2_ret SET x = t1_ret.x+2 FROM t1_ret WHERE t2_ret.id=t1_ret.id RETURNING after.x, before.x;
 x  | x 
----+---
  4 | 1
 10 | 4
(2 rows)

UPDATE t2_ret SET x = t1_ret.x*2 FROM t1_ret WHERE t2_ret.id=t1_ret.id RETURNING after.*, before.*;
 id | x  | y | z  | id | x  | y | z  
----+----+---+----+----+----+---+----
  1 |  4 | 2 |  6 |  1 |  4 | 2 |  6
  2 | 16 | 5 | 15 |  2 | 10 | 5 | 15
(2 rows)

DROP TABLE t1_ret;
DROP TABLE t2_ret;
